### PR TITLE
Add git:unlock command

### PR DESCRIPTION
**Pull request by Akirtovskis**
Related issue : https://github.com/dokku/dokku/issues/4458 

To solve the cloning issue : 
- added git:unlock command
- called within git:sync so on every next build clone directory is removed and build continues 
- added error handling if no `tty` is present (?) 
- removed cloned directory check from `fn-git-clone` as we now have check for that in `git:sync`

Questions that I still have : 
- when trying to run `git:unlock app` separately from vm it is not working and I see `git:unlock` is not a dokku command. Is 
there something obviuos I missed in command creation flow? 
- is this good error message for `no tty` cases? 
**dokku_log_fail "Missing tty, try running ssh -t which will allocate new tty for you"**

### CODE REVIEW

**Comment by josegonzalez** [comment]
The unlock shell script needs to be executable.

I would just error with a message telling the user to call git:unlock if no tty is present.

**Comment by Akirtovskis** [comment]
@josegonzalez Thanks for feedbacks. Fixed both issues and successfully tested `git:unlock` functionality on local.

File: plugins/git/internal-functions [hunk 1]
```diff
@@ -193,6 +193,20 @@ cmd-git-sync() {
 
   verify_app_name "$APP"
 
+  if [[ -d "$DOKKU_LIB_ROOT/data/git/$APP" ]]; then
+    if has_tty eq 0; then
+      cmd-git-unlock "$APP"
```
Comment (josegonzalez) [inline plugins/git/internal-functions:198]: We should _ask_ for confirmation before attempting to delete anything (thats the point of me asking you to check for a tty lol).

File: plugins/git/internal-functions [hunk 2]
```diff
@@ -193,6 +193,20 @@ cmd-git-sync() {
...
+      if [[ -d "$DOKKU_LIB_ROOT/data/git/$APP" ]]; then
+        dokku_log_fail "Failed to delete existing clone folder"
+        exit 15
+      fi
+    else
+      dokku_log_fail "Missing tty, please run git:unlock command"
```
Comment (josegonzalez) [inline plugins/git/internal-functions:204]: Should just be "Run 'git:unlock' to remove the existing temporary clone"

File: plugins/git/internal-functions [hunk 3]
```diff
@@ -193,6 +193,20 @@ cmd-git-sync() {
...
+    fi
+  else
+    dokku_log_info1 "No need to unlock app, carrying on with build"
```
Comment (josegonzalez) [inline plugins/git/internal-functions:207]: No need for this else block.

File: plugins/git/internal-functions [hunk 4]
```diff
@@ -215,6 +229,26 @@ cmd-git-sync() {
...
+  if [[ -d "$APP_CLONE_ROOT" ]]; then
+    dokku_log_info1 "About to delete $APP_CLONE_ROOT"
+    chown dokku:dokku -R "$APP_CLONE_ROOT"
+    chmod 755 "$APP_CLONE_ROOT"
+    rm -rf "$APP_CLONE_ROOT"
+    dokku_log_info1 "Successfully unlocked $APP"
+  else
+    dokku_log_info1 "No clone folder exists app already unlocked"
+  fi
+
```
Comment (josegonzalez) [inline plugins/git/internal-functions:249]: Remove the extra newline please.

File: plugins/git/internal-functions [hunk 5]
```diff
@@ -215,6 +229,26 @@ cmd-git-sync() {
...
+  local cmd="git:unlock"
+  [[ "$1" == "$cmd" ]] && shift 1
+  declare APP="$1"
+
+  local APP_CLONE_ROOT="$DOKKU_LIB_ROOT/data/git/$APP"
+
+  if [[ -d "$APP_CLONE_ROOT" ]]; then
+    dokku_log_info1 "About to delete $APP_CLONE_ROOT"
+    chown dokku:dokku -R "$APP_CLONE_ROOT"
+    chmod 755 "$APP_CLONE_ROOT"
```
Comment (josegonzalez) [inline plugins/git/internal-functions:243]: Do we need the chown/chmod?

**Review by josegonzalez — COMMENTED** [review]
Can you also write a test? Should just mock creating the `clone` directory.

The test might fail on gitlab because they don't support tty, so no need to check for the inline unlock, just want to know that the regular `git:unlock` command does what we expect.

File: plugins/git/internal-functions (hunk 1 above)
Comment (Akirtovskis) [reply plugins/git/internal-functions:198]: Ok, I will try to make this work and same for tests.

File: plugins/git/internal-functions (hunk 5 above)
Comment (Akirtovskis) [reply plugins/git/internal-functions:241]: Double checking this now. I added commands because at some point was facing `permission-denied` errors.

Comment (Akirtovskis) [reply plugins/git/internal-functions]: We do not, removed

**Comment by Akirtovskis** [comment]
@josegonzalez add the prompt and also --force flag for cases where it would make sense to bypass the prompt. 

Added tests also, but when I was running them they are failing with : 
```
(git) git:unlock [succes]
   (from function `global_setup' in file test_helper.bash, line 23,
    from function `setup' in test file git_6.bats, line 6)
     `global_setup' failed with status 127
   /Users/arturs/arturs/dev/dokku/tests/unit/test_helper.bash: line 23: free: command not found
   /Users/arturs/arturs/dev/dokku/tests/unit/test_helper.bash: line 216: dokku: command not found
   /Users/arturs/arturs/dev/dokku/tests/unit/test_helper.bash: line 37: dokku: command not found
```

Same is happening on other git test files as well and I am not sure how to fix that.

**Comment by josegonzalez** [comment]
Tests cant be run on OSX, they need to run in a Linux VM (the included one is fine).

**Comment by Akirtovskis** [comment]
@josegonzalez `git:unlock` command works as expected now and I even managed to make it cleaner by moving out the `fn-git-remove-clone-folder` to a separate function. 

But I would still appreciate some help with tests and fixing the lint issue.

**Comment by josegonzalez** [comment]
Gonna take over the PR from here.

**Pull request merged**